src/net/liveliness.c

```diff
@@ -0,0 +1,173 @@
+    ret_token->_zn = _z_session_rc_clone_as_weak(zn);
+    return ret;
+}
+
+z_result_t _z_undeclare_liveliness_token(_z_liveliness_token_t *token) {
+    if (token == NULL || _Z_RC_IS_NULL(&token->_zn)) {
+        return _Z_ERR_ENTITY_UNKNOWN;
+    }
+
+    z_result_t ret;
+
+    _z_liveliness_unregister_token(_Z_RC_IN_VAL(&token->_zn), token->_id);
+
+    _z_declaration_t declaration = _z_make_undecl_token(token->_id, &token->_key);
+    _z_network_message_t n_msg = _z_n_msg_make_declare(declaration, false, 0);
+    ret = _z_send_n_msg(_Z_RC_IN_VAL(&token->_zn), &n_msg, Z_RELIABILITY_RELIABLE, Z_CONGESTION_CONTROL_BLOCK);
+    _z_n_msg_clear(&n_msg);
+
+    return ret;
+}
+
+/**************** Liveliness Subscriber ****************/
+
+#if Z_FEATURE_SUBSCRIPTION == 1
+_z_subscriber_t _z_declare_liveliness_subscriber(const _z_session_rc_t *zn, _z_keyexpr_t keyexpr,
+                                                 _z_closure_sample_callback_t callback, _z_drop_handler_t dropper,
+                                                 bool history, void *arg) {
+    _z_subscription_t s;
+    s._id = _z_get_entity_id(_Z_RC_IN_VAL(zn));
+    s._key_id = keyexpr._id;
+    s._key = _z_get_expanded_key_from_key(_Z_RC_IN_VAL(zn), &keyexpr);
+    s._callback = callback;
+    s._dropper = dropper;
+    s._arg = arg;
+
+    _z_subscriber_t ret = _z_subscriber_null();
+    // Register subscription, stored at session-level, do not drop it by the end of this function.
+    _z_subscription_rc_t *sp_s =
+        _z_register_subscription(_Z_RC_IN_VAL(zn), _Z_SUBSCRIBER_KIND_LIVELINESS_SUBSCRIBER, &s);
+    if (sp_s == NULL) {
+        _z_subscriber_clear(&ret);
+        return ret;
+    }
+    // Build the declare message to send on the wire
+    uint8_t mode = history ? (_Z_INTEREST_FLAG_CURRENT | _Z_INTEREST_FLAG_FUTURE) : _Z_INTEREST_FLAG_FUTURE;
+    _z_interest_t interest = _z_make_interest(
+        &keyexpr, s._id, _Z_INTEREST_FLAG_KEYEXPRS | _Z_INTEREST_FLAG_TOKENS | _Z_INTEREST_FLAG_RESTRICTED | mode);
+
+    _z_network_message_t n_msg = _z_n_msg_make_interest(interest);
+    if (_z_send_n_msg(_Z_RC_IN_VAL(zn), &n_msg, Z_RELIABILITY_RELIABLE, Z_CONGESTION_CONTROL_BLOCK) != _Z_RES_OK) {
+        _z_unregister_subscription(_Z_RC_IN_VAL(zn), _Z_SUBSCRIBER_KIND_LIVELINESS_SUBSCRIBER, sp_s);
+        _z_subscriber_clear(&ret);
+        return ret;
+    }
+    _z_n_msg_clear(&n_msg);
+
+    ret._entity_id = s._id;
+    ret._zn = _z_session_rc_clone_as_weak(zn);
+    return ret;
+}
+
+z_result_t _z_undeclare_liveliness_subscriber(_z_subscriber_t *sub) {
+    if (sub == NULL || _Z_RC_IS_NULL(&sub->_zn)) {
+        return _Z_ERR_ENTITY_UNKNOWN;
+    }
+
+    _z_subscription_rc_t *s =
+        _z_get_subscription_by_id(_Z_RC_IN_VAL(&sub->_zn), _Z_SUBSCRIBER_KIND_LIVELINESS_SUBSCRIBER, sub->_entity_id);
+    if (s == NULL) {
+        return _Z_ERR_ENTITY_UNKNOWN;
+    }
+
+    _z_interest_t interest = _z_make_interest(NULL, s->_val->_id, 0);
```

Comment:
Since we have a `_z_make_interest_final` function, it would be good to use it here.